crates/ruff_linter/src/settings/types.rs

```diff
@@ -635,6 +635,78 @@ impl Deref for CompiledPerFileIgnoreList {
     }
 }
 
+// This struct and its `new` implementation are adapted directly from `PerFileIgnore`, minus the
+// `negated` field
+#[derive(Debug, Clone)]
+pub struct PerFileVersion {
+    pub(crate) basename: String,
+    pub(crate) absolute: PathBuf,
+    pub(crate) version: ast::PythonVersion,
+}
+
+impl PerFileVersion {
+    pub fn new(pattern: String, version: ast::PythonVersion, project_root: Option<&Path>) -> Self {
+        let absolute = match project_root {
+            Some(project_root) => fs::normalize_path_to(&pattern, project_root),
+            None => fs::normalize_path(&pattern),
+        };
+
+        Self {
+            basename: pattern,
+            absolute,
+            version,
+        }
+    }
+}
+
+#[derive(Debug, Clone, CacheKey)]
+pub struct CompiledPerFileVersion {
+    pub absolute_matcher: GlobMatcher,
+    pub basename_matcher: GlobMatcher,
+    pub version: ast::PythonVersion,
+}
+
+#[derive(CacheKey, Clone, Debug, Default)]
+pub struct CompiledPerFileVersionList {
+    versions: Vec<CompiledPerFileVersion>,
+}
+
+impl CompiledPerFileVersionList {
+    /// Given a list of patterns, create a `GlobSet`.
+    pub fn resolve(per_file_ignores: Vec<PerFileVersion>) -> Result<Self> {
+        let versions: Result<Vec<_>> = per_file_ignores
+            .into_iter()
+            .map(|per_file_version| {
+                // Construct absolute path matcher.
+                let absolute_matcher =
+                    Glob::new(&per_file_version.absolute.to_string_lossy())?.compile_matcher();
+
+                // Construct basename matcher.
+                let basename_matcher = Glob::new(&per_file_version.basename)?.compile_matcher();
+
```

Comment:
We should use `.context` or `with_context` or the error messages will be very unhelpful (e.g. invalid glob, good luck figuring out which glob).

We may also want to use `context` or `with_context` in `from_options`